pennylane/_qutrit_device.py

```diff
@@ -0,0 +1,390 @@
+        of wires specified.
+
+        Args:
+            num_wires (int): the number wires
+            dtype=np.uint32 (type): the data type of the arrays to use
+
+        Returns:
+            array[int]: the sampled basis states
+        """
+        # A slower, but less memory intensive method
+        basis_states_generator = itertools.product((0, 1, 2), repeat=num_wires)
+        return np.fromiter(itertools.chain(*basis_states_generator), dtype=dtype).reshape(
+            -1, num_wires
+        )
+
+    @staticmethod
+    def states_to_ternary(samples, num_wires, dtype=np.int64):
+        """Convert basis states from base 10 to ternary representation.
+
+        This is an auxiliary method to the generate_samples method.
+
+        Args:
+            samples (array[int]): samples of basis states in base 10 representation
+            num_wires (int): the number of qutrits
+            dtype (type): Type of the internal integer array to be used. Can be
+                important to specify for large systems for memory allocation
+                purposes.
+
+        Returns:
+            array[int]: basis states in ternary representation
+        """
+        ternary_arr = []
+        for sample in samples:
+            num = []
+            for _ in range(num_wires):
+                sample, r = divmod(sample, 3)
+                num.append(r)
+
+            ternary_arr.append(num[::-1])
+
+        return np.array(ternary_arr, dtype=dtype)
+
+    def density_matrix(self, wires):
+        """Returns the reduced density matrix prior to measurement.
+
+        .. note::
+
+            Only state vector simulators support this property. Please see the
+            plugin documentation for more details.
+        """
+        # TODO: Add density matrix support. Currently, qml.math is hard-coded to work only with qubit states,
+        # so it needs to be updated to be able to handle calculations for qutrits before this method can be
+        # implemented.
+        raise NotImplementedError
+
+    def estimate_probability(self, wires=None, shot_range=None, bin_size=None):
+        """Return the estimated probability of each computational basis state
+        using the generated samples.
+
+        Args:
+            wires (Iterable[Number, str], Number, str, Wires): wires to calculate
+                marginal probabilities for. Wires not provided are traced out of the system.
+            shot_range (tuple[int]): 2-tuple of integers specifying the range of samples
+                to use. If not specified, all samples are used.
+            bin_size (int): Divides the shot range into bins of size ``bin_size``, and
+                returns the measurement statistic separately over each bin. If not
+                provided, the entire shot range is treated as a single bin.
+
+        Returns:
+            array[float]: list of the probabilities
+        """
+
+        wires = wires or self.wires
```

Comment:
Aside from changing 2->3 in a few places, this function is identical to the qubit case as well, right?

Reply:
Yep.